lib/Containers/AtomicList.h

```diff
@@ -0,0 +1,274 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2025 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Business Source License 1.1 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+#include <vector>
+#include <type_traits>
+
+namespace arangodb {
+
+// In this file we implement a high performance append-only bounded
+// list. The list is bounded in that one can specify a limit on the used
+// memory and older entries are automatically freed. Appending is fast
+// in nearly all cases (two atomic operations). If an older chunk has
+// to be freed, deallocation of the old list entries must be performed.
+// With the exception of this relatively rare operation, the class is
+// lock-free. This is used to keep the most recent API calls and to be
+// able to deliver them via some API.
+
+// The class AtomicList is a simple lock-free implementation of a singly
+// linked list. One can only prepend new items and get a snapshot in form
+// of a raw pointer to a Node. With this one can traverse the list but
+// one must not free it.
+template<typename T>
+class AtomicList {
+ public:
+  // Note that Nodes use bare pointers, since AtomicList guards the allocation
+  // of the whole list.
+  struct Node {
+    T _data;
+    Node* _next;
+
+    Node(const T& value) : _data(value), _next(nullptr) {}
+    Node* next() { return _next; }
+  };
+
+ private:
+  std::atomic<Node*> _head;
+
+ public:
+  AtomicList() : _head(nullptr) {}
+
+  // It is *not* safe to destruct the AtomicList whilst other threads are
+  // still prepending! The user of the class has to ensure that this is
+  // done properly!
+  ~AtomicList() {
+    Node* n = _head.load();
+    while (n != nullptr) {
+      Node* next = n->next();
+      delete n;
+      n = next;
+    }
+    _head.store(nullptr);
+  }
+
+  void prepend(T&& value) noexcept {
+    // Note that this
```

Comment:
I think something is missing in this comment or it is obsolete.

Reply:
Removed.